<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Tutorial:exonize_analysis - exonize</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
        <link href="../extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Tutorial:exonize_analysis";
        var mkdocs_page_input_path = "tutorial.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> exonize
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Overview</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../usage/">Usage</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../output/">Output</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../exonize_analysis/">exonize_analysis</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Tutorial:exonize_analysis</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#example-human-y-chromosome">Example: Human Y chromosome</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#data-representation">Data representation</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#plotting">Plotting</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">exonize</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Tutorial:exonize_analysis</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/msarrias/exonize/edit/master/docs/tutorial.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="exonize_analysis-module-tutorial"><code>exonize_analysis</code> module tutorial<a class="headerlink" href="#exonize_analysis-module-tutorial" title="Permanent link">&para;</a></h1>
<p>This guide will walk you through the basics of using the <code>exonize_analysis</code> module for parsing and analyzing the output data generated by <code>exonize</code>.</p>
<h2 id="example-human-y-chromosome">Example: Human Y chromosome<a class="headerlink" href="#example-human-y-chromosome" title="Permanent link">&para;</a></h2>
<h3 id="data-representation">Data representation<a class="headerlink" href="#data-representation" title="Permanent link">&para;</a></h3>
<p><strong>Step 1: Import the <code>exonize_analysis</code> module</strong></p>
<p><div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">exonize.exonize_analysis</span> <span class="kn">import</span> <span class="n">ExpansionsContainer</span>
</code></pre></div>
<strong>Step 2: Set the path to the exonize results database</strong></p>
<p>For this example, we’ll use the results database generated for the Y chromosome in the usage <a href="https://msarrias.github.io/exonize/usage/#example-human-y-chromosome">example</a>.</p>
<p><div class="highlight"><pre><span></span><code><span class="n">db_path</span><span class="o">=</span><span class="s1">&#39;Homo_sapiens_chrom_Y_exonize/Homo_sapiens_chrom_Y_results.db&#39;</span>
</code></pre></div>
<strong>Step 3: Create an <a href="https://msarrias.github.io/exonize/exonize_analysis/#exonize.exonize_analysis.ExpansionsContainer"><code>ExpansionsContainer</code></a> object</strong> </p>
<p>Initialize the <a href="https://msarrias.github.io/exonize/exonize_analysis/#exonize.exonize_analysis.ExpansionsContainer"><code>ExpansionsContainer</code></a> object using the specified database. This class is particularly useful for performing specific gene lookups and accessing their expansions. Its attributes consist of <a href="https://msarrias.github.io/exonize/exonize_analysis/#exonize.exonize_analysis.Gene"><code>Gene</code></a> objects composed by <a href="https://msarrias.github.io/exonize/exonize_analysis/#exonize.exonize_analysis.Expansion"><code>Expansion</code></a> graphs.</p>
<div class="highlight"><pre><span></span><code><span class="n">ychrom_expansions</span> <span class="o">=</span> <span class="n">ExpansionsContainer</span><span class="p">(</span><span class="n">exonize_db_path</span><span class="o">=</span><span class="n">db_path</span><span class="p">)</span>
</code></pre></div>
<p>Now let’s check the number of genes identified with duplication events:</p>
<div class="highlight"><pre><span></span><code><span class="nb">len</span><span class="p">(</span><span class="n">ychrom_expansions</span><span class="p">)</span>
<span class="mi">15</span>
</code></pre></div>
<p><strong>Step 4: List genes with exon duplications</strong></p>
<p>Display some of the gene IDs where exon duplications were found:</p>
<div class="highlight"><pre><span></span><code><span class="nb">print</span><span class="p">(</span><span class="n">ychrom_expansions</span><span class="o">.</span><span class="n">genes</span><span class="p">[:</span><span class="mi">4</span><span class="p">])</span>
<span class="p">[</span><span class="s1">&#39;gene:ENSG00000292357&#39;</span><span class="p">,</span> <span class="s1">&#39;gene:ENSG00000165246&#39;</span><span class="p">,</span> <span class="s1">&#39;gene:ENSG00000012817&#39;</span><span class="p">,</span> <span class="s1">&#39;gene:ENSG00000244395&#39;</span><span class="p">]</span>
</code></pre></div>
<p><strong>Step 5: Examine the DAZ1 gene</strong></p>
<p>Let’s take a closer look at the <a href="https://en.wikipedia.org/wiki/DAZ1">DAZ1</a> gene (Ensembl ID: <em>gene:ENSG00000205916</em>).
<div class="highlight"><pre><span></span><code><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">]</span>
<span class="o">&lt;</span><span class="n">Gene</span> <span class="n">gene</span><span class="p">:</span><span class="n">ENSG00000205916</span> <span class="k">with</span> <span class="mi">6</span> <span class="n">expansions</span> <span class="p">(</span><span class="n">iterable</span> <span class="n">of</span> <span class="n">expansion</span> <span class="n">graphs</span><span class="p">)</span><span class="o">&gt;</span>
</code></pre></div>
Note that the <a href="https://msarrias.github.io/exonize/exonize_analysis/#exonize.exonize_analysis.Gene"><code>Gene</code></a> object is an iterable of <a href="https://msarrias.github.io/exonize/exonize_analysis/#exonize.exonize_analysis.Expansion"><code>Expansion</code></a> objects, where each <a href="https://msarrias.github.io/exonize/exonize_analysis/#exonize.exonize_analysis.Expansion"><code>Expansion</code></a> is a NetworkX graph structure encoding the expansions. You can find more details about the attributes of these graph structures in the <a href="https://networkx.org/documentation/stable/reference/classes/graph.html">NetworkX documentation</a>.</p>
<p>DAZ1 has 6 expansions:</p>
<div class="highlight"><pre><span></span><code><span class="nb">len</span><span class="p">(</span><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">])</span>
<span class="mi">6</span>
</code></pre></div>
<p><strong>Step 6: Inspect the largest expansion</strong></p>
<div class="highlight"><pre><span></span><code><span class="n">expansion_id</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="nb">len</span><span class="p">)</span><span class="o">.</span><span class="n">id</span>
</code></pre></div>
<p>Check the size:</p>
<div class="highlight"><pre><span></span><code><span class="nb">len</span><span class="p">(</span><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">][</span><span class="n">expansion_id</span><span class="p">])</span>
<span class="mi">18</span>
</code></pre></div>
<p>This output indicates that DAZ1 contains 18 copies of an exon. Now, let’s look at the mode of these duplications:</p>
<p><div class="highlight"><pre><span></span><code><span class="n">mode</span> <span class="o">=</span> <span class="p">[</span><span class="n">attrib</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;mode&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">node</span><span class="p">,</span> <span class="n">attrib</span> <span class="ow">in</span> <span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">][</span><span class="n">expansion_id</span><span class="p">]</span><span class="o">.</span><span class="n">nodes</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span>
<span class="n">mode_counts</span> <span class="o">=</span> <span class="p">{</span><span class="n">t</span><span class="p">:</span> <span class="n">mode</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">mode</span><span class="p">)}</span>
<span class="n">mode_counts</span>
<span class="p">{</span><span class="s1">&#39;FULL&#39;</span><span class="p">:</span> <span class="mi">15</span><span class="p">,</span> <span class="s1">&#39;INTRONIC&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">}</span>
</code></pre></div>
This tells us that there are matches between 15 exons and 3 intronic regions. </p>
<h3 id="plotting">Plotting<a class="headerlink" href="#plotting" title="Permanent link">&para;</a></h3>
<p><strong>Step 1: Visualize the expansion</strong></p>
<p>Now, let’s visualize the expansion graph for the DAZ1 gene. Nodes represent coordinates and edges indicate matches found between them:</p>
<div class="highlight"><pre><span></span><code><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">draw_expansions_multigraph</span><span class="p">(</span><span class="n">expansion_id</span><span class="o">=</span><span class="n">expansion_id</span><span class="p">)</span>
</code></pre></div>
<div style="text-align: center;">
    <a href="https://github.com/msarrias/exonize/blob/main/figures/DAZ1_expansion1.png" target="_blank">
        <img src="https://github.com/msarrias/exonize/raw/main/figures/DAZ1_expansion1.png" alt="DAZ1_expansion1"  width="400">
    </a>
</div>

<p><strong>Step 2: Visualize the gene structure</strong></p>
<p>The <code>draw_gene_structure</code> method uses the <a href="https://edinburgh-genome-foundry.github.io/DnaFeaturesViewer/">dna_features_viewer</a> library to illustrate the positions of expansion events within the gene, along with the coding exons it comprises.</p>
<p>First, we need to parse the gene hierarchy dictionary, which can be found in the Exonize output directory. This step is necessary, since information about the arrangement of the exons within the genes is not stored in the output database.</p>
<p><div class="highlight"><pre><span></span><code><span class="n">ychrom_expansions</span><span class="o">.</span><span class="n">parse_gene_hierarchy_dictionary</span><span class="p">(</span>
    <span class="n">gene_hierarchy_dictionary_path</span><span class="o">=</span><span class="s1">&#39;Homo_sapiens_chrom_Y_exonize/Homo_sapiens_chrom_Y_gene_hierarchy.pkl&#39;</span>
<span class="p">)</span>
</code></pre></div>
Let's now plot the gene's features:</p>
<div class="highlight"><pre><span></span><code><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">draw_gene_structure</span><span class="p">(</span><span class="n">expansion_id</span><span class="o">=</span><span class="n">expansion_id</span><span class="p">)</span>
</code></pre></div>
<div style="text-align: center;">
    <a href="https://github.com/msarrias/exonize/blob/main/figures/gene_structure.png" target="_blank">
        <img src="https://github.com/msarrias/exonize/raw/main/figures/gene_structure.png" alt="gene_structure"  width="500">
    </a>
</div>

<p>The dark bars indicate the locations of the gene's coding sequences, while the light bars highlight the locations of the expansion events, colored according to the event mode.</p>
<p><strong>Step 3: Visualize the full expansion</strong></p>
<p>To visualize the full expansion, and matches between tandem exon pairs set the <code>full_expansion</code> and <code>color_tandem_pair_edges</code> parameters to <code>True</code>.</p>
<div class="highlight"><pre><span></span><code><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">draw_expansions_multigraph</span><span class="p">(</span>
    <span class="n">expansion_id</span><span class="o">=</span><span class="n">expansion_id</span><span class="p">,</span>
    <span class="n">full_expansion</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">color_tandem_pair_edges</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</code></pre></div>
<div style="text-align: center;">
    <a href="https://github.com/msarrias/exonize/blob/main/figures/DAZ1_expansion1_full.png" target="_blank">
        <img src="https://github.com/msarrias/exonize/raw/main/figures/DAZ1_expansion1_full.png" alt="DAZ1_expansion1_full"  width="400">
    </a>
</div>

<p>We can see that the expansion graph is composed of 15 tandemly duplicated exons.</p>
<p><strong>Step 4: Visualize the match graph for DAZ1</strong></p>
<p>You can also display all expansions associated with DAZ1 by ommiting the <code>expansion_id</code> parameter:</p>
<div class="highlight"><pre><span></span><code><span class="n">ychrom_expansions</span><span class="p">[</span><span class="s1">&#39;gene:ENSG00000205916&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">draw_expansions_multigraph</span><span class="p">(</span>
    <span class="n">full_expansion</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">color_tandem_pair_edges</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</code></pre></div>
<div style="text-align: center;">
    <a href="https://github.com/msarrias/exonize/blob/main/figures/DAZ1_expansion1_full_allexp.png" target="_blank"  width="700">
        <img src="https://github.com/msarrias/exonize/raw/main/figures/DAZ1_expansion1_full_allexp.png" alt="DAZ1_expansion1_full_allexp" >
    </a>
</div>

<p>We can see that all 6 events found in step 5 refer to full exon duplications, where only the first expansion is found to be composed by tandem pairs.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../exonize_analysis/" class="btn btn-neutral float-left" title="exonize_analysis"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/msarrias/exonize" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../exonize_analysis/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
