```
███████╗██╗  ██╗ ██████╗ ███╗   ██╗██╗███████╗███████╗
██╔════╝╚██╗██╔╝██╔═══██╗████╗  ██║██║╚══███╔╝██╔════╝
█████╗   ╚███╔╝ ██║   ██║██╔██╗ ██║██║  ███╔╝ █████╗
██╔══╝   ██╔██╗ ██║   ██║██║╚██╗██║██║ ███╔╝  ██╔══╝
███████╗██╔╝ ██╗╚██████╔╝██║ ╚████║██║███████╗███████╗
╚══════╝╚═╝  ╚═╝ ╚═════╝ ╚═╝  ╚═══╝╚═╝╚══════╝╚══════╝
Marina Herrera Sarrias, Department of Mathematics, Stockholm University
Christopher Wheat, Department of Zoology, Stockholm University
Liam M. Longo, Earth-Life Science Institute (ELSI), Institute of Science Tokyo
Lars Arvestad, Department of Mathematics, Stockholm University
```

## Welcome!
`exonize` is an open-source command-line tool and [Python package](https://pypi.org/project/exonize/) for identifying and classifying coding exon duplications in annotated genomes. `exonize` identifies full exon duplications using local and global alignment methods and implements a graph-based framework to handle clusters of exons formed by repetitive duplication events. In addition, `exonize` categorizes the interdependence between duplicated exons (or groups of exons) across transcripts. For data parsing and downstream analysis, the `exonize_analysis` module is available for Python notebooks.

## Documentation

Please see the [Documentation](https://msarrias.github.io/exonize/) for a full user guide and an introductory tutorial to the `exonize_analysis` module.

## Support

If you need help with `exonize`, contact us! To report a bug or request a new feature, open an **[Issue](https://github.com/msarrias/exonize/issues)** on the `exonize` repo.

## Citation

If you use `exonize` in a publication, please cite:
```
TBA
```

