`exonize_analysis` module tutorial
============

This guide will walk you through the basics of using the `exonize_analysis` module for parsing and analyzing the output data generated by `exonize`.

### Example: Human Y chromosome

**Step 1: Import the `exonize_analysis` Module**

```python
from exonize import exonize_analysis as exonize
```
**Step 2: Set the Path to the Exonize Results Database**

For this example, we’ll use the results database generated for the Y chromosome in the usage [example](https://msarrias.github.io/exonize/usage/#example-human-y-chromosome).

```python
db_path='Homo_sapiens_chrom_Y_exonize/Homo_sapiens_chrom_Y_results.db'
```
**Step 3: Create a `GenomeExpansions` Object**

Initialize the `GenomeExpansions` object using the specified database.

```python
ychrom_expansions = exonize.GenomeExpansions(exonize_db_path=db_path)
```

Now let’s check the number of genes identified with duplication events:

```python
>>> len(ychrom_expansions)
15
```

**Step 4: List Genes with Exon Duplications**

Display some of the gene IDs where exon duplications were found:

```python
>>> print(ychrom_expansions.genes[-4:])
['gene:ENSG00000188120', 'gene:ENSG00000165246', 'gene:ENSG00000205916', 'gene:ENSG00000205944']
```

**Step 5: Examine the DAZ1 Gene**

Let’s take a closer look at the [DAZ1](https://en.wikipedia.org/wiki/DAZ1) gene (Ensembl id _gene:ENSG00000205916_). notice that the Gene object will return an interable of `Expansion` objects.

```python
>>> ychrom_expansions['gene:ENSG00000205916']
<Gene gene:ENSG00000205916 with 6 expansions (iterable of expansion graphs)>
```

DAZ1 has 6 expansions:

```python
>>> len(ychrom_expansions['gene:ENSG00000205916'])
6
```

**Step 6: Inspect Expansion #1**

Check the size of expansion #1:

```python
>>> len(ychrom_expansions['gene:ENSG00000205916'][1])
18
```

This output shows that an exon in DAZ1 duplicated 18 times. Now, let’s look at the types of these duplications:

```python
>>> types = [attrib.get('type') for node, attrib in ychrom_expansions['gene:ENSG00000205916'][1].nodes(data=True)]
>>> type_counts = {t: types.count(t) for t in set(types)}
>>> type_counts
{'FULL': 15, 'INTRONIC': 3}
```
This tells us that there are matches between 15 exons and 3 intronic regions. 

**Step 7: Visualize the Expansion #1 Graph**

Now, let’s visualize the graph for the DAZ1 gene, where nodes represent coordinates and edges indicate matches found by `exonize`:

```python
>>> ychrom_expansions['gene:ENSG00000205916'].draw_expansions_multigraph(
    expansion_id=1,
    figure_size=(10,7)
)
```

<div style="text-align: center;">
    <a href="https://github.com/msarrias/exonize/blob/main/figures/DAZ1_expansion1.png" target="_blank">
        <img src="https://github.com/msarrias/exonize/raw/main/figures/DAZ1_expansion1.png" alt="DAZ1_expansion1"  width="400">
    </a>
</div>

**Step 8: View the Full Expansion**

To visualize the full expansion, set the `full_expansion` parameter to `True`.

```python
>>> ychrom_expansions['gene:ENSG00000205916'].draw_expansions_multigraph(
    expansion_id=1,
    full_expansion=True,
    figure_size=(10,7)
)
```

<div style="text-align: center;">
    <a href="https://github.com/msarrias/exonize/blob/main/figures/DAZ1_expansion1_full.png" target="_blank">
        <img src="https://github.com/msarrias/exonize/raw/main/figures/DAZ1_expansion1_full.png" alt="DAZ1_expansion1_full"  width="400">
    </a>
</div>

**Step 9: Visualize All Expansions for DAZ1**

You can also display all expansions associated with DAZ1 by ommiting the `expansion_id` parameter:

```python
>>> ychrom_expansions['gene:ENSG00000205916'].draw_expansions_multigraph(
    full_expansion=True,
    figure_size=(10,7)
)
```

<div style="text-align: center;">
    <a href="https://github.com/msarrias/exonize/blob/main/figures/DAZ1_expansion1_full_allexp.png" target="_blank">
        <img src="https://github.com/msarrias/exonize/raw/main/figures/DAZ1_expansion1_full_allexp.png" alt="DAZ1_expansion1_full_allexp" >
    </a>
</div>

We can see that all 6 events found in step 5 refer to full exon duplications.